feat: (payments v3) Implement temporal specific error handling in connectors

laouji opened the pull request:
[ENG-1371](https://linear.app/formance/issue/ENG-1371/payments-v3x-implement-temporal-specific-errors-for-retries-or)

This PR only covers plugins that use the httpwrapper. I'll get Ayden and Stripe in a different PR since this one is big enough.

Dav-14 commented on internal/connectors/plugins/public/bankingcircle/balances.go:
```diff
@@ -31,20 +33,20 @@ func (p Plugin) fetchNextBalances(ctx context.Context, req models.FetchNextBalan
 		// circle response. We will use the current time instead.
 		// lastTransactionTimestamp, err := time.Parse("2006-01-02T15:04:05.999999999+00:00", balance.LastTransactionTimestamp)
 		// if err != nil {
-		// 	return models.FetchNextBalancesResponse{}, fmt.Errorf("failed to parse opening date: %w", err)
+		// 	return models.FetchNextBalancesResponse{}, models.NewPluginError(fmt.Errorf("failed to parse opening date: %w", err))
```
Is all those comment still needed ?

laouji replied:
Ah good question. Was this intended to be commented out or do we need to override the timestamp @paul-nicolas ?

paul-nicolas replied:
This is why we do lastTransactionTimestamp := time.Now().UTC() instead of parsing the balance transaction timestamp. But yeah we can keep only the comment and not the commented code I think

laouji replied:
removed

Dav-14 commented on internal/connectors/plugins/public/mangopay/bank_account_creation.go:
```diff
@@ -58,7 +60,9 @@ func (p Plugin) createBankAccount(ctx context.Context, req models.CreateBankAcco
 			if req.BankAccount.AccountNumber == nil {
-				return models.CreateBankAccountResponse{}, fmt.Errorf("missing account number in bank account metadata")
+				return models.CreateBankAccountResponse{}, models.NewPluginError(
+					fmt.Errorf("missing account number in bank account metadata"),
```
Can this be a normalized error ?

laouji replied:
done

Dav-14 commented on internal/connectors/plugins/public/mangopay/bank_account_creation.go:
```diff
@@ -73,12 +77,14 @@ func (p Plugin) createBankAccount(ctx context.Context, req models.CreateBankAcco
 			if req.BankAccount.AccountNumber == nil {
-				return models.CreateBankAccountResponse{}, fmt.Errorf("missing account number in bank account metadata")
+				return models.CreateBankAccountResponse{}, models.NewPluginError(
+					fmt.Errorf("missing account number in bank account metadata"),
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/bank_account_creation.go:
```diff
@@ -93,12 +99,14 @@ func (p Plugin) createBankAccount(ctx context.Context, req models.CreateBankAcco
 			if req.BankAccount.AccountNumber == nil {
-				return models.CreateBankAccountResponse{}, fmt.Errorf("missing account number in bank account metadata")
+				return models.CreateBankAccountResponse{}, models.NewPluginError(
+					fmt.Errorf("missing account number in bank account metadata"),
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/bank_account_creation.go:
```diff
@@ -112,12 +120,14 @@ func (p Plugin) createBankAccount(ctx context.Context, req models.CreateBankAcco
 			if req.BankAccount.AccountNumber == nil {
-				return models.CreateBankAccountResponse{}, fmt.Errorf("missing account number in bank account metadata")
+				return models.CreateBankAccountResponse{}, models.NewPluginError(
+					fmt.Errorf("missing account number in bank account metadata"),
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/bank_accounts.go:
```diff
@@ -130,16 +128,11 @@ func (c *Client) createBankAccount(ctx context.Context, endpoint string, req any
-		return nil, err
+	statusCode, err := c.httpClient.Do(httpReq, &bankAccount, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create bank account, got code %d: %w", statusCode, err)
```
can this used the sharedError pkg ?

laouji replied:
Are you referring to a package in particular, or do you just mean to create a shared package which helps format such errors?

laouji replied:
I updated the mangopay client to use `go-libs/errorsutils` to handle the status code as you suggested.

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/bank_accounts.go:
```diff
@@ -167,13 +160,9 @@ func (c *Client) GetBankAccounts(ctx context.Context, userID string, page, pageS
-		return nil, err
+	statusCode, err := c.httpClient.Do(req, &bankAccounts, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get bank accounts, got code %d: %w", statusCode, err)
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/payin.go:
```diff
@@ -41,13 +39,9 @@ func (c *Client) GetPayin(ctx context.Context, payinID string) (*PayinResponse,
-		return nil, err
+	statusCode, err := c.httpClient.Do(req, &payinResponse, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get payin response got code %d: %w", statusCode, err)
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/payout.go:
```diff
@@ -62,16 +60,11 @@ func (c *Client) InitiatePayout(ctx context.Context, payoutRequest *PayoutReques
-		// Never retry payout initiation
-		return nil, err
+	statusCode, err := c.httpClient.Do(req, &payoutResponse, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to initiate payout, got code %d: %w", statusCode, err)
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/payout.go:
```diff
@@ -88,13 +81,9 @@ func (c *Client) GetPayout(ctx context.Context, payoutID string) (*PayoutRespons
-		return nil, err
+	statusCode, err := c.httpClient.Do(req, &payoutResponse, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get payout response got code %d: %w", statusCode, err)
```
same

Dav-14 commented on internal/connectors/plugins/public/mangopay/client/refund.go:
```diff
@@ -42,13 +40,9 @@ func (c *Client) GetRefund(ctx context.Context, refundID string) (*Refund, error
-		return nil, err
+	statusCode, err := c.httpClient.Do(req, &refund, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get refund, got code %d: %w", statusCode, err)
```
same

laouji commented:
@gfyrag If you don't like type assertions let's discuss what some other options are?

My general idea was to keep logic related to Temporal as close to the activity definitions as possible. I wanted to avoid having any logic related to temporal implemented in the client packages themselves since they should just be concerned with communicating with the various APIs.

What I set out to achieve in this PR was:
1) API clients can return `error` interface without knowledge of temporal
2) a small subset of those errors, upon being unwrapped can be converted into temporal "NoRetry" errors

I suppose instead of a having an abstracted `models.PluginError` type I could create a generic function like
```
func toTemporalError(err error) *temporal.ApplicationError {
  if errors.Is(err, SomeErr) {
  } else if...
}
```

Is that really nicer than a quick type assertion though?


> return nil, fmt.Errorf("failed to make payout, status code %d: %w", statusCode, err)
> The original error is lost. It prevent a client from being capable to unwrap the errors.

How so? the `%w` embeds the error in a way that allows it to be unwrapped by `errors.Is()` later. Actually I am relying on that mechanic to be able to detect http 4xx statuses.

laouji commented:
@gfyrag I removed the type assertion in favour of a simple function that uses `errors.Is()` to determine non-retryable errors as we discussed. I also added some regression tests which should fail if the plugin functions are refactored in a way which alters the errors they return.

paul-nicolas commented on internal/connectors/plugins/public/bankingcircle/client/accounts.go:
```diff
@@ -69,14 +67,10 @@ func (c *Client) GetAccounts(ctx context.Context, page int, pageSize int, fromOp
-		// TODO(polo): retryable errors
-		return nil, fmt.Errorf("received status code %d for get accounts", statusCode)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get accounts, status code %d: %w", statusCode, err)
```
I don't think you want to loose the httpwrapper error, an errors.Wrap is needed here to keep the custom error from httpwrapper

laouji replied:
Actually no, it's not needed since Go 1.13. As I discussed with @gfyrag yesterday, the github.com/pkg/errors is mostly obsolete (note that it is archived) as the only functionality it provides that `fmt.Errorf()` doesn't is the stack trace. Since we aren't actually making use of a stack trace downstream, we can start using `fmt.Errorf()` everywhere and try to remove the dependency of github.com/pkg/errors from this repo going forward.

Here's a good overview if you need some history of the changes since Go 1.13:
https://stackoverflow.com/questions/61933650/whats-the-difference-between-errors-wrapf-errors-errorf-and-fmt-errorf

paul-nicolas commented on internal/connectors/plugins/public/bankingcircle/client/accounts.go:
```diff
@@ -97,12 +91,8 @@ func (c *Client) GetAccount(ctx context.Context, accountID string) (*Account, er
-	case httpwrapper.ErrStatusCodeUnexpected:
-		// TODO(polo): retryable errors
-		return nil, fmt.Errorf("received status code %d for get account", statusCode)
+	if err != nil {
```
same here, we need a Wrap

paul-nicolas commented:
I'am adding a comment here: everytime you do a `c.httpClient.Do(req, &res, nil)` which returns specific errors depending on the status code, you should do an `errors.Wrap` to keep the specific error instead of creating a new one

paul-nicolas commented:
> I'am adding a comment here: everytime you do a `c.httpClient.Do(req, &res, nil)` which returns specific errors depending on the status code, you should do an `errors.Wrap` to keep the specific error instead of creating a new one

My bad, now the idiomatic way to do it is with the %w !
